lib/ui/src/containers/nav.js

```diff
@@ -95,6 +96,64 @@ const createMenu = memoize(1)((api, shortcutKeys, isFullscreen, showPanel, showN
   },
 ]);
 
+const recursiveRemove = (stories, id) => {
+  const node = stories[id];
+  // eslint-disable-next-line no-param-reassign
+  delete stories[id];
+  if (node.parent) {
+    const parent = stories[node.parent];
+    parent.children = parent.children.filter(child => child !== id);
+    if (!parent.children.length) {
+      recursiveRemove(stories, parent.id);
+    }
+  }
+};
+
+export const removeDocsOnlyStories = stories => {
+  const filtered = cloneDeep(stories);
+  Object.values(stories).forEach(({ id, parameters }) => {
+    if (parameters && parameters.docsOnly) {
+      recursiveRemove(filtered, id);
+    }
+  });
+  return filtered;
+};
+
+export const flattenDocsOnlyStories = stories => {
+  const filtered = cloneDeep(stories);
+  Object.values(stories).forEach(story => {
+    const { id, isLeaf, parent, parameters } = story;
+    if (isLeaf && parameters && parameters.docsOnly) {
+      if (!filtered[parent].isLeaf) {
+        filtered[parent].isLeaf = true;
+        filtered[parent].id = id;
+        delete filtered[parent].children;
+      }
+      delete filtered[id];
+    }
+  });
+  return filtered;
+};
+
+export const removeLeafStories = stories => {
+  const filtered = cloneDeep(stories);
+  Object.values(stories).forEach(story => {
+    const { id, isLeaf, parent } = story;
+    if (isLeaf) {
+      if (!filtered[parent].isLeaf) {
+        filtered[parent].isLeaf = true;
```

Comment:
This logic seems wrong. It is possible to have both leaves and non leaves as children. For instance

```
GroupA
|-- GroupB
|      \-- Story1
\-- Story 2
```

I think there are some more complex examples in the code base, eg. in https://github.com/storybookjs/storybook/blob/next/lib/ui/src/components/sidebar/treeview/treeview.mockdata.js#L1.

I think to be correct, you'd need to delete `story` from `parent.children`, *then* set it to `isLeaf` if it is empty

Reply:
Can we just do this (without clone) in a fairly simple way:

```js
return stories.filter(s => !s.isLeaf).map(({ children, ...story }) => {
  const newChildren = children.filter(s => !s.isLeaf);
  return newChildren.length === 0 ? {...story, isLeaf: true } : {...story, children: newChildren};
});
```